### PR TITLE
ENH: Plot cross-correlations and auto/cross-correlation matrix

### PR DESCRIPTION
**Pull request by aglebov**
- New function `plot_ccf` to plot cross-correlations by lag between two variables, analogous to `plot_acf`.
- New function `plot_accf` to plot a matrix of auto- and cross-correlation plots between several variables.

### CODE REVIEW

**Comment by aglebov** [comment]
This pull request assumes that #8782 is accepted.

**Comment by josef-pkt** [comment]
I just saw that VAR has also a plot_acorr method (actually 2 different versions, one in results and one in process)

https://stats.stackexchange.com/questions/616550/statsmodels-var-plot-acorr-amount-of-plots/616558

**Comment by aglebov** [comment]
Thanks - yes, here is the output from VARResults.plot_acorr() for the same
data:

[image: VAR.png]
This is workable. The disadvantages I can see are:

   - one has to instantiate a VAR model and fit it just to produce the
   plots, although one might not be interested in a VAR(p) model itself,
   - the implementation does not go through statsmodels.tsa.stattools.acf()
   or ccf() to calculate the values, so the functionality might diverge.

Do you think it makes sense to have model-independent utilities to
calculate and plot ACF/CCF in
statsmodels.tsa.stattools/statsmodels.graphics.tsaplots and reimplement
VARResults.plot_acorr() to use those?

**Comment by josef-pkt** [comment]
> Do you think it makes sense to have model-independent utilities to
calculate and plot ACF/CCF in
statsmodels.tsa.stattools/statsmodels.graphics.tsaplots and reimplement
VARResults.plot_acorr() to use those?

I'm not sure.
VAR plot_acorr should be improved. 
avoiding code duplication would also be useful.

What I don't know is whether acf/ccf should be computed in an array (vectorized) or in a loop, or have both versions.
The VAR plot function uses precomputed acorr matrix.
Your version computes pairwise acf/ccf when needed in the plot loop. I don't know whether it makes sense to precompute the acorr array outside of plotting.

**Comment by aglebov** [comment]
> Can you rebase on main, and preferable squash the ccf commits, (i.e. partially squash into 2 or 3 commits)?

Done, there was just one new commit on top of the ccf changes from #8782.

Please let me know if you want to change the names of the `plot_accf` - I don't have any great ideas for the name.

File: statsmodels/graphics/tsaplots.py [hunk 1]
```diff
@@ -367,6 +368,266 @@ def plot_pacf(
...
+    zero : bool, optional
+        Flag indicating whether to include the 0-lag autocorrelations
+        (which are always equal to 1). Default is True.
+    auto_ylims : bool, optional
+        If True, adjusts automatically the vertical axis limits
+        to correlation values.
+    bartlett_confint : bool, default True
+        If True, use Bartlett's formula to calculate confidence intervals
+        in autocorrelation plots. See the description of ``plot_acf`` for
+        details.
```
Comment (josef-pkt) [inline statsmodels/graphics/tsaplots.py:545]: should mention that bartlett_confint does not affect ccf confint.

File: statsmodels/graphics/tsaplots.py [hunk 2]
```diff
@@ -367,6 +368,266 @@ def plot_pacf(
...
+        fft=False,
+        missing="none",
+        zero=True,
+        auto_ylims=False,
+        bartlett_confint=True,
+        vlines_kwargs=None,
+        **kwargs,
+):
+    """
+    Plot matrix of auto- and cross-correlations
```
Comment (josef-pkt) [inline statsmodels/graphics/tsaplots.py:501]: plot grid

we used "grid" for correlation plots instead of matrix

**Comment by josef-pkt** [comment]
change function name to `plot_accf_grid`
for correlation plot and similar, we added "_grid" postfix.

I think using "accf" in name is fine. 
(just using ccf or acf would be also ok, but accf is more explicit even though it's not standard)

**Comment by aglebov** [comment]
> change function name to `plot_accf_grid`

Done

**Comment by josef-pkt** [comment]
I have not looked at grid plots in some time.

how are varnames handled? do we get names on the x and y axes of the grid?

**Comment by josef-pkt** [comment]
Something like the seaborn plot grids https://seaborn.pydata.org/tutorial/axis_grids.html#plotting-pairwise-data-relationships
would be nice to have.

But not really necessary for this PR.

File: statsmodels/graphics/tsaplots.py [hunk 3]
```diff
@@ -367,6 +368,266 @@ def plot_pacf(
...
+    gs = fig.add_gridspec(m, m)
+
+    if _is_using_pandas(x, None):
+        varnames = list(x.columns)
+
+        def get_var(i):
+            return x.iloc[:, i]
+    else:
+        varnames = [f'x[{i}]' for i in range(m)]
+
```
[the remark this answers is not in the document]
Comment (aglebov) [reply statsmodels/graphics/tsaplots.py:595]: Done

**Comment by aglebov** [comment]
> Something like the seaborn plot grids https://seaborn.pydata.org/tutorial/axis_grids.html#plotting-pairwise-data-relationships would be nice to have.

One difference here is that ACF/CCF plots show lags on the x axis and correlations on the y axis (unlike pair plots where the values of the variables are plotted), so it might be confusing to label the axes with variable names.

As currently implemented, the output looks like this for a pandas dataframe:

image

If a numpy array is passed, the titles are less useful:

image

But with the new `varnames` argument this could be overridden to produce the same output as for pandas.

The inspiration for the plot titles comes from R - here is what the `acf` function shows for the same data:

image

Happy to change it in whatever way you like.

Please also note that I've changed the default value of the `bartlett_formula` argument from `True` to `False`, so that the confidence intervals for ACF and CCF are consistent by default (since `bartlett_formula=True` only affects ACF).

**Comment by josef-pkt** [comment]
> One difference here is that ACF/CCF plots show lags on the x axis and correlations on the y axis (unlike pair plots where the values of the variables are plotted), so it might be confusing to label axes with variable names.

I did not think about that. The current plot grid is fine then.

One more clarification:
Looking at the plot, it was not immediately obvious to me which is the main variable and which is the lagged variable in rows and columns of the grid, especially if the names are connected by "&"

You could add a sentence to the `Return` section, stating that
Plots on the grid show the cross-correlation of the row variable with the lags of the column variable.

**Comment by aglebov** [comment]
> You could add a sentence to the `Return` section, stating that
> Plots on the grid show the cross-correlation of the row variable with the lags of the column variable.

Done.

I've kept the display as shown below - please let me know if you want to change anything:

image

**Comment by aglebov** [comment]
Actually, let me also fix one more thing: the confidence interval shading does not cover all of the lags displayed for CCF.

**Comment by aglebov** [comment]
> Actually, let me also fix one more thing: the confidence interval shading does not cover all of the lags displayed for CCF.

Done, here is the output after the fix:

image

**Comment by josef-pkt** [comment]
LGTM, thanks for the PRs

merging

**Comment by nsingh216** [comment]
Hello,
Could anyone point me to the documentation for these new functions?
Thanks!

**Comment by aglebov** [comment]
> Hello, Could anyone point me to the documentation for these new functions? Thanks!

Hey, good spot, the functions seem to be missing from the generated docs - I'll try to add them.

In the meantime, you can try looking at the docstrings in the code:
https://github.com/statsmodels/statsmodels/blob/1b0113a81527b9a4a90a6c562bcc32869dd052e2/statsmodels/graphics/tsaplots.py#L372
https://github.com/statsmodels/statsmodels/blob/1b0113a81527b9a4a90a6c562bcc32869dd052e2/statsmodels/graphics/tsaplots.py#L487

**Comment by aglebov** [comment]
Pull request #9299 adds these functions to the generated docs.

**Comment by aglebov** [comment]
@nsingh216 in case you are still interested, the docs are now available:
https://www.statsmodels.org/dev/generated/statsmodels.graphics.tsaplots.plot_ccf.html
https://www.statsmodels.org/dev/generated/statsmodels.graphics.tsaplots.plot_accf_grid.html

**Pull request merged**